go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,306 @@
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type Log struct {
+	// Example: val{int64_val: 5017}, val{int64_val: 1003}
+	EntityValue []*types.Value
+
+	FeatureValues   []*types.Value
+	FeatureStatuses []serving.FieldStatus
+	EventTimestamps []*timestamppb.Timestamp
+	RequestContext  map[string]*types.RepeatedValue
+}
+
+type MemoryBuffer struct {
+	featureService *model.FeatureService
+	logs           []*Log
+}
+
+type LoggingService struct {
+	memoryBuffer      *MemoryBuffer
+	logChannel        chan *Log
+	fs                *feast.FeatureStore
+	offlineLogStorage OfflineLogStorage
+}
+
+func NewLoggingService(fs *feast.FeatureStore, logChannelCapacity int, featureServiceName string, enableLogging bool) (*LoggingService, error) {
+	var featureService *model.FeatureService = nil
+	var err error
+	if enableLogging {
+		featureService, err = fs.GetFeatureService(featureServiceName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	loggingService := &LoggingService{
+		logChannel: make(chan *Log, logChannelCapacity),
+		memoryBuffer: &MemoryBuffer{
+			logs:           make([]*Log, 0),
+			featureService: featureService,
+		},
+		fs: fs,
+	}
+
+	if !enableLogging || fs == nil {
+		loggingService.offlineLogStorage = nil
+	} else {
+		offlineLogStorage, err := NewOfflineStore(fs.GetRepoConfig())
+		loggingService.offlineLogStorage = offlineLogStorage
+
+		if err != nil {
+			return nil, err
+		}
+		// Start goroutine to process logs
+		go loggingService.processLogs()
+	}
+	return loggingService, nil
+}
+
+func (s *LoggingService) EmitLog(l *Log) error {
+	select {
+	case s.logChannel <- l:
+		return nil
+	case <-time.After(20 * time.Millisecond):
+		return fmt.Errorf("could not add to log channel with capacity %d. Operation timed out. Current log channel length is %d", cap(s.logChannel), len(s.logChannel))
+	}
+}
+
+func (s *LoggingService) processLogs() {
+	// start a periodic flush
+	// TODO(kevjumba): set param so users can configure flushing duration
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		s.ProcessMemoryBuffer(ticker)
+	}
+}
+
+// Select that eitheringests new logs that are added to the logging channel, one at a time to add
+// to the in memory buffer or flushes all of them synchronously to the OfflineStorage on a time interval.
+func (s *LoggingService) ProcessMemoryBuffer(t *time.Ticker) {
+	select {
+	case t := <-t.C:
+		s.flushLogsToOfflineStorage(t)
+	case new_log := <-s.logChannel:
+		log.Printf("Adding %s to memory.\n", new_log.FeatureValues)
+		s.memoryBuffer.logs = append(s.memoryBuffer.logs, new_log)
+	}
+}
+
+// Acquires the logging schema from the feature service, converts the memory buffer array of rows of logs and flushes
+// them to the offline storage.
+func (s *LoggingService) flushLogsToOfflineStorage(t time.Time) error {
+	log.Printf("Flushing buffer to offline storage with channel length: %d\n at time: "+t.String(), len(s.memoryBuffer.logs))
```

Comment:
`log.Printf("... "+t.String()` if format is used, let's use format for all parameters

Reply:
Actually I'll remove this.